fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -0,0 +1,217 @@
+                readVal = Tuple.fromBytes(bytes).getLong(0); // surely flushed
+                assertEquals(readVal, i);
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void assertLoopThreadsValues() {
+        try (FDBRecordContext context = fdb.openContext()) {
+            for (int loop = 0; loop < loopCount; loop++) {
+                final AgilityContext agilityContext = AgilityContext.factory(context, false);
+                for (int thread = 0; thread < threadCount; thread++) {
+                    int i = loop * thread;
+                    byte[] key = Tuple.from(500, i).pack();
+                    byte[] val = Tuple.from(i).pack();
+                    agilityContext.get(key).thenApply(bytes -> {
+                        assertEquals(val, bytes);
+                        final long existingVal = Tuple.fromBytes(bytes).getLong(0);
+                        assertEquals(i, existingVal);
+                        return null;
+                    });
+                }
+            }
+            context.commit();
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testAgilityContextConcurrent(boolean useAgile)  throws ExecutionException, InterruptedException {
+        try (FDBRecordContext context = fdb.openContext()) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, useAgile);
+            testAgilityContextConcurrentSingleObject(agilityContext, true);
+            context.commit();
+        }
+        assertLoopThreadsValues();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testAgilityContextConcurrentNonExplicitCommits(boolean useAgile) throws ExecutionException, InterruptedException {
+        for (int sizeQuota : new int[] {1, 2, 7, 21, 100, 10000}) {
+            final RecordLayerPropertyStorage.Builder insertProps = RecordLayerPropertyStorage.newBuilder()
+                    .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_SIZE_QUOTA, sizeQuota);
+
+            try (FDBRecordContext context = openContext(insertProps)) {
+                final AgilityContext agilityContext = AgilityContext.factory(context, useAgile);
+                testAgilityContextConcurrentSingleObject(agilityContext, false);
+                context.commit();
+            }
+        }
+        assertLoopThreadsValues();
+    }
+
+    void testAgilityContextOneLongWrite(int loopCount, int sizeLimit, int timeLimit) {
+        final RecordLayerPropertyStorage.Builder insertProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_SIZE_QUOTA, sizeLimit)
+                .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_TIME_QUOTA, timeLimit);
+
+        String ShelSilverstein =
+                "Oh, if you’re a bird, be an early bird\n" +
```

Comment:
This quote is from a work still under copyright, it might fall under fair-use but it would be better to pick text that's definitely in the public domain (before 1900 should be ok)

Reply:
✅